### ISSUE TITLE
ExistValidator, targetRelation does not work when owner and related active record use different ActiveQuery implementation

### ISSUE THREAD

**Issue by phrakon** [yiisoft/yii2#18822]
### What steps will reproduce the problem?

```php
class Unit extends \yii2tech\filedb\ActiveRecord
{
}

class Product extends \yii\db\ActiveRecord
{
    public function getUnit()
    {
        return $this->hasOne(Unit::class, ['id' => 'unit_id']);
    }

    public function rules()
    {
        return [
            [['unit_id'], 'exist', 'targetRelation' => 'unit'],
        ];
    }
}

$product = new Product(['unit_id'=>1]);
$product->save();
```

### What is the expected result?

No errors.

### What do you get instead?

Calling unknown method: yii\db\Connection::getQueryProcessor()

It tries to use Product connection instead of Unit connection to read data.

It worked ok on 2.0.42.1

### Additional info

| Q                | A
| ---------------- | ---
| Yii version      | 2.0.?
| PHP version      | 7.4.21

| Operating system | Ubuntu 20.04

**Comment by phrakon** [comment]
Looks like
https://github.com/yiisoft/yii2/commit/1474e2d61a46b36d824a4211d89d42ed9268619c#diff-a183cabc1fe0fc2db443994c8400c1a806fdaad8c9eb5fc189bedeb5d4e31e3d
broke previous functionalty

**Comment by bizley** [comment]
@WinterSilence please take a look at this.

**Comment by WinterSilence** [comment]
@bizley 
problem not in current PR. You can revert changes and try run given example at `$forceMasterDb = false` - no effect. To fix it method must use `$relationQuery->primaryModel::getDb()` instead `$model::getDb()`, but it's broke logic of `$forceMasterDb`

**Comment by phrakon** [comment]
looks like if we use

`$exists = $connection->useMaster([$relationQuery, 'exists']);`

current $connection passed to exists() method of $relationQuery

if we use 
```
            $exists = $connection->useMaster(function() use ($relationQuery) {
                return $relationQuery->exists();
            });
```
passed null

**Comment by WinterSilence** [comment]
@phrakon you right, `$forceMasterDb` ignored in any case

**Comment by phrakon** [comment]
Anyway, my case should use forceMasterDb=false.
I think your commit fixed an old error. It should be
```
 $exists = $connection->useMaster(function($db) use ($relationQuery) {
                return $relationQuery->exists($db);
            });
```
useMaster passes current connection to related. I was wrong to use forceMasterDb=true, and the mistake confused me.

**Comment by WinterSilence** [comment]
@phrakon 
> I think your commit fixed an old error. 

yes. but I'm didn't thinked problem was "deeper".

**Comment by phrakon** [comment]
The case two errors are getting code work :) My opinion, they should make forceMasterDb equals false by default

**Comment by WinterSilence** [comment]
@phrakon **your** code.
@bizley Еxplore problem:
- "master DB" is connection using in `$model` i.e. `$model::getDB()`
- `$relationQuery->exists();` use connection `$relationQuery->primaryModel::getDb()`
- If `$forceMasterDb` is true, then must be `$relationQuery->exists($model::getDB());`
I'm right?

**Comment by bizley** [comment]
I need to take a closer look at this since I didn't expect that your update would cause problem, it looked all right. Let me get back to you with this.

**Comment by WinterSilence** [comment]
@bizley I already said that problem was even before my PR, I'm only fix it. it's happened because your tests don't cover it. `$forceMasterDb` must be `false` by default.

**Comment by bizley** [comment]
@phrakon can you confirm this? You said it was ok on 2.0.42.1

**Comment by WinterSilence** [comment]
@bizley what does @phrakon have to do with it? what will change if I'm set `$forceMasterDb = false`? Nothing? (:

**Comment by phrakon** [comment]
I solved it by adding forceMasterDb = false. I wrote above, it looks like @WinterSilence fixed an old bug. My opinion is that you should set forceMasterDb = false by default several year ago (when the feature was implemented).
It broke my code due to the fact that I did not study the validator well enough, and an error in the framework did not allow it to be immediately identified. Those on the one hand, it would be necessary to accept the fix, on the other hand, it can break the code of someone like me.

**Comment by WinterSilence** [comment]
after deep exploring i'm find answer - method have invalid logic and use `$model::getDB()` instead `$relationQuery->primaryModel::getDb()`:
~~~php
private function checkTargetRelationExistence($model, $attribute)
{
    $exists = false;
    $method = 'get' . ucfirst($this->targetRelation);
    /** @var ActiveQuery $relationQuery */
    $relationQuery = $model->{$method}();
    if (!$relationQuery instanceof ActiveQuery) {
       throw new \yii\base\InvalidValueException(
          'Method ' . get_class($model) . '::' .  $method . ' must return instance of yii\db\ActiveQuery'
       );
    }

    if ($this->filter instanceof \Closure) {
        call_user_func($this->filter, $relationQuery);
    } elseif ($this->filter !== null) {
        $relationQuery->andWhere($this->filter);
    }

    if ($this->forceMasterDb && $relationQuery->primaryModel instanceof ActiveRecord) {
        $connection = $relationQuery->primaryModel::getDb();
        $exists = $connection->useMaster([$relationQuery, 'exists']);
    } else {
        $exists = $relationQuery->exists();
    }

    if (!$exists) {
        $this->addError($model, $attribute, $this->message);
    }
}
~~~

Also, I'm found many LSP BC in class `ActiveRecord`, but we skip this now..

**Comment by WinterSilence** [comment]
"master DB" is [yii\db\Connection::$master](https://www.yiiframework.com/doc/api/2.0/yii-db-connection#$master-detail)

**Comment by bizley** [comment]
I can confirm that even though DB connection from the second model is used the called method comes from the first model.

**Comment by bizley** [comment]
@WinterSilence that won't work - primaryModel points to the parent model so it won't use relation model's query object. I've got some idea to just fetch the relation model and use that but we will lose benefits of just calling `exists()`.

**Comment by bizley** [comment]
Yes, `$relationQuery->modelClass` points to the relation model. Current implementation is to call `exists()` on the primary model relation method and that is the problem.

**Comment by bizley** [comment]
I'm updating the title of the issue because it's wrong.

**Comment by bizley** [comment]
@samdark this shouldn't be closed as I understand, even after rollback (which is weird anyway because WinterSilence said that this is not a problem in his update - I agree with that - and yet he rolled it back) the problem is still here. I've tested it on a previous implementation.

**Comment by terabytesoftw** [comment]
Closing as no longer reproducible on current Yii2 master (May 2026).

I verified the reported setup using a `yii\db\ActiveRecord` owner model with a `targetRelation` pointing to a `yii2tech\filedb\ActiveRecord` model. Validation succeeds for an existing related record and fails normally for a missing one, with both `forceMasterDb=true` and `forceMasterDb=false`.

The original exception is still reproducible only with the old callback form that passed the owner DB connection into the relation query. The current implementation executes the relation query without passing that connection, so custom ActiveQuery implementations such as `yii2tech\filedb\ActiveQuery` use their own connection as expected.

if this still happens, please provide a fresh reproduction against Yii2 `2.0.55` or current master.
